Headers/AppKit/NSCollectionView.h

```diff
@@ -44,47 +58,401 @@ enum
+
+/**
+ * Returns the set of indexPaths that should change.
+ */ 
+- (NSSet *) collectionView: (NSCollectionView *)collectionView
+shouldSelectItemsAtIndexPaths: (NSSet *)indexPaths;
+
+/**
+ * Returns the set of NSIndexPath objects that should be deselected given the proposed indexPaths.
+ */
+- (NSSet *) collectionView: (NSCollectionView *)collectionView shouldDeselectItemsAtIndexPaths: (NSSet *)indexPaths;
+
+/**
+ * Called to inform the delegate of those items that were selected.
+ */
+- (void) collectionView: (NSCollectionView *)collectionView didSelectItemsAtIndexPaths: (NSSet *)indexPaths;
+
+/**
+ * Called to inform the delegate of those items that were deselected.
+ */
+- (void) collectionView: (NSCollectionView *)collectionView didDeselectItemsAtIndexPaths: (NSSet *)indexPaths;
+
+/* Display Notification */
+
+/**
+ * Called to inform the delegate that the item representing the object at indexPath will be displayed.
+ */
+- (void) collectionView: (NSCollectionView *)collectionView
+        willDisplayItem: (NSCollectionViewItem *)item
+        forRepresentedObjectAtIndexPath: (NSIndexPath *)indexPath;
+
+/**
+ * Called to inform the delegate that the supplementary view for the elementKind will be displayed 
+ * at indexPath.
+ */
+- (void) collectionView: (NSCollectionView *)collectionView
+         willDisplaySupplementaryView: (NSView *)view
+         forElementKind: (NSCollectionViewSupplementaryElementKind)elementKind
+            atIndexPath: (NSIndexPath *)indexPath;
+
+/**
+ * Called to inform the delegate that the collectionView will end display of item for the object at indexPath.
+ */
+- (void) collectionView: (NSCollectionView *)collectionView
+   didEndDisplayingItem: (NSCollectionViewItem *)item
+   forRepresentedObjectAtIndexPath: (NSIndexPath *)indexPath;
+
+/**
+ * Called to inform the delegate that the collectionView will end display of supplementaryView for the 
+ * object at indexPath.
+ */
+- (void) collectionView: (NSCollectionView *)collectionView
+         didEndDisplayingSupplementaryView: (NSView *)view
+       forElementOfKind: (NSCollectionViewSupplementaryElementKind)elementKind
+            atIndexPath: (NSIndexPath *)indexPath;
+
+/* Layout Transition Support */
+
+/**
+ * Called to inform the delegate that the collectionView is transitioning from the old to the new
+ * layout indicated.
+ */
+- (NSCollectionViewTransitionLayout *) collectionView: (NSCollectionView *)collectionView
+                         transitionLayoutForOldLayout: (NSCollectionViewLayout *)fromLayout
+                                            newLayout: (NSCollectionViewLayout *)toLayout;
+#endif
+@end
 
 APPKIT_EXPORT_CLASS
 @interface NSCollectionView : NSView //<NSDraggingDestination, NSDraggingSource>
 {
+  // Content
   NSArray *_content;
-  IBOutlet NSCollectionViewItem *itemPrototype;
   NSMutableArray *_items;
-  
-  BOOL _allowsMultipleSelection;
-  BOOL _isSelectable;
   NSIndexSet *_selectionIndexes;
```

Comment:
The don't need to be.  I thought we had covered this.   Basically with these changes we, essentially,  have two separate versions of this class in ONE file.   One handles things using an array controller and indexes and the other uses indexPaths.   So I don't think they need to be kept in sync...  I will test this, but I doubt there is any relationship between the two.

Reply:
```
Enabled
2023-02-04 10:20:49.810566-0500 NSCollectionViewLayout_test[41075:3382853] indexPaths = _NSCollectionViewMutableIndexPathSet {1:(4)}
2023-02-04 10:20:49.810717-0500 NSCollectionViewLayout_test[41075:3382853] *** Assertion failure in NSIndexSet *_NSIndexSetFromCollectionViewIndexPaths(NSSet<NSIndexPath *> *)_block_invoke(), NSCollectionView.m:9000
2023-02-04 10:20:49.811025-0500 NSCollectionViewLayout_test[41075:3382853] [General] NSIndexPath(s) with section != 0
2023-02-04 10:20:49.816050-0500 NSCollectionViewLayout_test[41075:3382853] [General] (
	0   CoreFoundation                      0x00007ff80b29640b __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x00007ff80ade2e25 objc_exception_throw + 48
	2   Foundation                          0x00007ff80c5b9067 -[NSMutableDictionary(NSMutableDictionary) initWithContentsOfFile:] + 0
	3   AppKit                              0x00007ff80edc5284
```

When NSIndexPaths are used, it apparently throws an exception when you even try to access the value of "selectionIndexes"... here is the code...

```

- (void)collectionView:(NSCollectionView *)collectionView didSelectItemsAtIndexPaths:(NSSet<NSIndexPath *> *)indexPaths
{
    NSLog(@"selectionIndexes: %@", [self.collectionView selectionIndexes]);
    NSLog(@"selectionIndexPaths: %@", [self.collectionView selectionIndexPaths]);
}
@end
```

Reply:
I just implemented the call for the "collectionView:didSelectItemsAtIndexPaths:" it just prints the values of both.   As I had mentioned, I don't think they need to be kept in sync.   If you want me to implement the exception, I will, but I think that is something we don't really need to do for GS.

Reply:
I don't see how the stack trace proves what you think. Why would Apple have a function for this conversion, when it is always bound to fail? But you were able to see your whole program and if you are sure these two variables are unrelated go ahead and implement it that way.

Reply:
I'm not sure if you have a copy of any version of macOS that contains the new enhancements. 

The original implementation of NSCollectionView only implemented one layout and kept track of things using NSIndexSet which merely tracks the the selected indices of the items not the indexPaths (which are different).   Additionally, the original implementation uses an NSArrayController and a content array. 

The enhancement which implements layouts and can support custom layouts tracks selections using a set of index paths and uses the delegate / data source approach.  

Can you figure out a way to reconcile index sets and index paths??  If so I would very much like to hear what your opinion is. Is there a compelling reason to keep them in sync?  I know this from looking at the header and playing with the implementation in the test which I wrong.  

Additionally, when using the array controller that exception that I showed you doesn't occur.

Reply:
> I don't see how the stack trace proves what you think. Why would Apple have a function for this conversion, when it is always bound to fail? But you were able to see your whole program and if you are sure these two variables are unrelated go ahead and implement it that way.

YOU are able to see my whole program too.  I have repeatedly stated where I put every single test I make every time I submit a PR.  Please review the history of this PR to find it as it has been stated at least one time.  

:/

Reply:
https://github.com/gcasa/NSCollectionViewLayout_test

Reply:
As discussed during the meeting and commented on below, Apple is, apparently, keeping these in sync.  The exception is from when the section is != 0, but does not appear otherwise.  I had assumed, incorrectly, that since it wasn't possible to keep them in sync in the general case that it wouldn't be done.  The Apple implementation issues an exception whereas the GNUstep code issues a log warning.   This difference is acceptable, I believe because the exception does not cause a crash and also because this functionality is entirely undocumented. GC

Reply:
I am resolving this for now, please resurrect it if you find any other issues.   I am hoping there aren't that many more things to review.   I am going to leave the fixes for the legacy functionality and the composite layout for separate PRs given that this PR is massive, and old, already.  GC